psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -908,4 +912,1372 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                        }
+                        touched = true;
+                        // if (thread == 0) timer_off("JK: GEMV");
+                    }
+                }
+            }
+        }  // End Shell Quartets
+
+        if (!touched) continue;
+
+        // => Stripe out <= //
+
+        // if (thread == 0) timer_on("JK: Atomic");
+        for (size_t ind = 0; ind < D.size(); ind++) {
+            double** JKTp = JKT[thread][ind]->pointer();
+            double** Jp = J[ind]->pointer();
+
+            double* J1p = JKTp[0L * max_task];
+            double* J2p = JKTp[1L * max_task];
+
+            // > J_PQ < //
+
+            for (int P2 = 0; P2 < nPtask; P2++) {
+                for (int Q2 = 0; Q2 < nQtask; Q2++) {
+                    int P = task_shells[P2start + P2];
+                    int Q = task_shells[Q2start + Q2];
+                    int Psize = primary_->shell(P).nfunction();
+                    int Qsize = primary_->shell(Q).nfunction();
+                    int Poff = primary_->shell(P).function_index();
+                    int Qoff = primary_->shell(Q).function_index();
+                    int Poff2 = task_offsets[P2 + P2start] - task_offsets[P2start];
+                    int Qoff2 = task_offsets[Q2 + Q2start] - task_offsets[Q2start];
+                    for (int p = 0; p < Psize; p++) {
+                        for (int q = 0; q < Qsize; q++) {
+#pragma omp atomic
+                            Jp[p + Poff][q + Qoff] += J1p[(p + Poff2) * dQsize + q + Qoff2];
+                        }
+                    }
+                }
+            }
+
+            // > J_RS < //
+
+            for (int R2 = 0; R2 < nRtask; R2++) {
+                for (int S2 = 0; S2 < nStask; S2++) {
+                    int R = task_shells[R2start + R2];
+                    int S = task_shells[S2start + S2];
+                    int Rsize = primary_->shell(R).nfunction();
+                    int Ssize = primary_->shell(S).nfunction();
+                    int Roff = primary_->shell(R).function_index();
+                    int Soff = primary_->shell(S).function_index();
+                    int Roff2 = task_offsets[R2 + R2start] - task_offsets[R2start];
+                    int Soff2 = task_offsets[S2 + S2start] - task_offsets[S2start];
+                    for (int r = 0; r < Rsize; r++) {
+                        for (int s = 0; s < Ssize; s++) {
+#pragma omp atomic
+                            Jp[r + Roff][s + Soff] += J2p[(r + Roff2) * dSsize + s + Soff2];
+                        }
+                    }
+                }
+            }
+
+        }  // End stripe out
+        // if (thread == 0) timer_off("JK: Atomic");
+
+    }  // End master task list
+
+    for (size_t ind = 0; ind < D.size(); ind++) {
+        J[ind]->scale(2.0);
+        J[ind]->hermitivitize();
+    }
+
+    if (bench_) {
+        auto mode = std::ostream::app;
+        auto printer = std::make_shared<PsiOutStream>("bench.dat", mode);
```

Comment:
A `shared_ptr` is completely unnecessary. Just `PsiOutStream printer("bench.dat", mode)` is preferred. Apply to all functions.